<a href="https://colab.research.google.com/github/JLMuehlbauer/Project-2-Classification/blob/main/Project_2_JacksonMuehlbauer.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Project 2
---
Jackson Muehlbauer

Date: 01/09/2023

In [1]:
# Imports (for now)
import pandas as pd
import numpy as np

In [7]:
# Mount Drive
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


## Dataset Option 1:
---
**Source:** https://www.kaggle.com/datasets/pschale/mlb-pitch-data-20152018/versions/2?resource=download

**Brief Description:**

This dataset represents MLB pitches across the 2015-2018 seasons. The features represent various metrics that describe the given pitch as well as the outcome of the pitch and the status of the at bat/game. 

In [9]:
path = '/content/drive/My Drive/Colab Notebooks/Raw Data/2019_pitches.csv'

df = pd.read_csv(path)
pd.set_option('display.max_columns', None)
df.head()


px    pz  start_speed  end_speed    spin_rate     spin_dir  break_angle  \
0  0.00  2.15         88.8       80.7  placeholder  placeholder         22.8   
1  0.34  2.31         89.9       81.8  placeholder  placeholder         22.8   
2 -0.05  2.03         85.7       79.6  placeholder  placeholder          9.6   
3  0.49  0.92         85.4       78.5  placeholder  placeholder         24.0   
4 -0.13  1.11         84.6       77.6  placeholder  placeholder         26.4   

   break_length  break_y     ax     ay     az  sz_bot  sz_top type_confidence  \
0           4.8     24.0  -8.47  28.90 -15.51    1.70    3.36     placeholder   
1           3.6     24.0  -7.10  28.85 -12.99    1.80    3.55     placeholder   
2           6.0     24.0   3.65  22.07 -22.64    1.59    3.55     placeholder   
3           7.2     24.0 -13.77  24.44 -25.74    1.74    3.55     placeholder   
4           8.4     24.0 -15.99  24.56 -28.36    1.83    3.59     placeholder   

    vx0     vy0   vz0       x    x0       y    y0    z0  pfx_x  pfx_z  nasty  \
0  5.28 -128.95 -6.89  116.97 -1.42  180.81  50.0  6.07  -5.07   9.98    NaN   
1  4.89 -130.54 -7.48  103.93 -1.02  176.34  50.0  6.20  -4.14  11.18    NaN   
2  2.33 -124.60 -5.98  118.86 -1.29  183.96  50.0  6.30   2.30   5.99    NaN   
3  7.83 -123.74 -6.78   98.15 -1.56  214.03  50.0  5.85  -8.87   4.14    NaN   
4  6.79 -122.67 -5.73  121.81 -1.57  208.77  50.0  5.89 -10.51   2.51    NaN   

          zone code type pitch_type  event_num  b_score         ab_id  \
0  placeholder    X    X         FF          5      0.0  2.019000e+09   
1  placeholder    C    C         FF          8      0.0  2.019000e+09   
2  placeholder    S    S         SL          9      0.0  2.019000e+09   
3  placeholder    B    B         CH         10      0.0  2.019000e+09   
4  placeholder    B    B         CH         11      0.0  2.019000e+09   

   b_count  s_count  outs  pitch_num  on_1b  on_2b  on_3b  
0      0.0      0.0   0.0        1.0    0.0    0.0    0.0  
1      0.0      0.0   1.0        1.0    0.0    0.0    0.0  
2      0.0      0.0   1.0        2.0    0.0    0.0    0.0  
3      0.0      1.0   1.0        3.0    0.0    0.0    0.0  
4      1.0      1.0   1.0        4.0    0.0    0.0    0.0

In [10]:
df.columns

Index(['px', 'pz', 'start_speed', 'end_speed', 'spin_rate', 'spin_dir',
       'break_angle', 'break_length', 'break_y', 'ax', 'ay', 'az', 'sz_bot',
       'sz_top', 'type_confidence', 'vx0', 'vy0', 'vz0', 'x', 'x0', 'y', 'y0',
       'z0', 'pfx_x', 'pfx_z', 'nasty', 'zone', 'code', 'type', 'pitch_type',
       'event_num', 'b_score', 'ab_id', 'b_count', 's_count', 'outs',
       'pitch_num', 'on_1b', 'on_2b', 'on_3b'],
      dtype='object')

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 728790 entries, 0 to 728789
Data columns (total 40 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   px               722161 non-null  float64
 1   pz               722161 non-null  float64
 2   start_speed      722161 non-null  float64
 3   end_speed        722161 non-null  float64
 4   spin_rate        722161 non-null  object 
 5   spin_dir         722161 non-null  object 
 6   break_angle      722161 non-null  float64
 7   break_length     722161 non-null  float64
 8   break_y          722161 non-null  float64
 9   ax               722161 non-null  float64
 10  ay               722161 non-null  float64
 11  az               722161 non-null  float64
 12  sz_bot           728790 non-null  float64
 13  sz_top           728790 non-null  float64
 14  type_confidence  722161 non-null  object 
 15  vx0              722161 non-null  float64
 16  vy0              722161 non-null  floa

In [8]:
df.shape

(728790, 40)

In [16]:
df['pitch_type'].value_counts()

FF    261485
SL    127680
CH     79939
CU     64166
FT     61044
SI     55361
FC     46339
KC     15324
FS     10538
KN       200
EP        49
FO        36
Name: pitch_type, dtype: int64

**What is the Target?**

I am interested in completing a classification type problem, thus, there are 2 options I could see as the target for this dataset: 
1. 'code', this feature represents the outcome of the pitch. For example, 'S' is a swinging strike.
2. 'pitch_type', this feature is the category that the pitch falls into. For example, 'FF' is four-seam fastball. 

If I were to move forward with this problem, I would want to build a model that predicts the 'pitch_type'. The purpose of this model would simply be to classify and display the predicted pitch type to a TV or live audience such that they could follow the game more closely. The reason I chose 'pitch_type' over 'code' is to significantly simplify the model... more to come.

**What does one row represent?**
One row represents a single thrown pitch in an MLB game.

**Is this a classification or regression problem?**

As stated earlier, I would treat this as a classification problem. 

**How many features does the data have?**

This dataset has 40 features (including the potential target). While this is a lot, by choosing to classify 'pitch_type', the following features can be dropped as they will have little to no influence on classifying what type of pitch was thrown (or they are empty):

['spin_rate', 'spin_dir', 'sz_bot', 'sz_top', 'type_confidence', 'nasty', 'zone', 'code', 'type', 'event_num', 'b_score', 'ab_id', 'b_count', 's_count', 'outs', 'pitch_num', 'on_1b', 'on_2b', 'on_3b', 'y0'] 

This brings the total features down to 19 with 1 target. 

**How many rows are in the dataset?**

There are 728790 rows in this data set. That said, a much small subset can easily be used for this type of problem

**What, if any, challenges do you foresee in cleaning, exploring, or modeling with this dataset?**

There are many features and many rows in this dataset (with some null values). It could require a fair amount of effort to clean this set. Additionally, the large number of rows and columns could increase the time for training and could potentially disqualify lazy models such as KNN. The last challenge is that there is no "positive" class, this is a multiclass problem, were each class is no more important than the other. This may make evaluating the model metrics harder to interpret.

## Dataset Option 2:
---
**Source:** https://www.kaggle.com/datasets/fedesoriano/stroke-prediction-dataset

**Brief Description:**

This dataset shows people who did and did not have a stroke along with various health related metrics.


In [12]:
path_2 = '/content/drive/My Drive/Colab Notebooks/Raw Data/healthcare-dataset-stroke-data.csv'
df_2 = pd.read_csv(path_2)
df_2.head()

id  gender   age  hypertension  heart_disease ever_married  \
0   9046    Male  67.0             0              1          Yes   
1  51676  Female  61.0             0              0          Yes   
2  31112    Male  80.0             0              1          Yes   
3  60182  Female  49.0             0              0          Yes   
4   1665  Female  79.0             1              0          Yes   

       work_type Residence_type  avg_glucose_level   bmi   smoking_status  \
0        Private          Urban             228.69  36.6  formerly smoked   
1  Self-employed          Rural             202.21   NaN     never smoked   
2        Private          Rural             105.92  32.5     never smoked   
3        Private          Urban             171.23  34.4           smokes   
4  Self-employed          Rural             174.12  24.0     never smoked   

   stroke  
0       1  
1       1  
2       1  
3       1  
4       1

In [13]:
df_2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5110 entries, 0 to 5109
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   id                 5110 non-null   int64  
 1   gender             5110 non-null   object 
 2   age                5110 non-null   float64
 3   hypertension       5110 non-null   int64  
 4   heart_disease      5110 non-null   int64  
 5   ever_married       5110 non-null   object 
 6   work_type          5110 non-null   object 
 7   Residence_type     5110 non-null   object 
 8   avg_glucose_level  5110 non-null   float64
 9   bmi                4909 non-null   float64
 10  smoking_status     5110 non-null   object 
 11  stroke             5110 non-null   int64  
dtypes: float64(3), int64(4), object(5)
memory usage: 479.2+ KB


In [14]:
df_2.shape

(5110, 12)

In [15]:
df_2['stroke'].value_counts()

0    4861
1     249
Name: stroke, dtype: int64

**What is the Target?**

The target is 'stroke'

**What does one row represent?**

Each row represents a people that did or did not have a stroke. 

**Is this a classification or regression problem?**

This as a classification problem as we would like to predict the likihood that someone has a stroke. 

**How many features does the data have?**

This data has 12 columns (11 features).

**How many rows are in the dataset?**

There are 5110 rows in this data set.

**What, if any, challenges do you foresee in cleaning, exploring, or modeling with this dataset?**

This dataset is fairly imbalanced. I suspect that a fair amount of work will be required to tune the model to reduce False Negatives. It may also require some effort to evaluate the performance of the model. 